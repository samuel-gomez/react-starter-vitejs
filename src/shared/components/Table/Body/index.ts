export { default } from './Body';
export type { TBody } from './Body';
