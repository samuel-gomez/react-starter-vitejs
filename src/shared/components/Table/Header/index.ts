export { default } from './Header';
export type { THeader } from './Header';
