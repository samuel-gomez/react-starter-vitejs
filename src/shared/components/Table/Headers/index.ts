export { default, type THeaders } from './Headers';
