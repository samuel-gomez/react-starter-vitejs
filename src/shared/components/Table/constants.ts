export const NONE = 'NONE';
export const ASCENDING = 1;
export const DESCENDING = -1;

const EORDER = {
  NONE: 'NONE',
  ASCENDING: -1,
  DESCENDING: 1,
};

export default EORDER;

export const DEFAULT_TABLE_ITEMS_TYPE = 'donn√©es';
