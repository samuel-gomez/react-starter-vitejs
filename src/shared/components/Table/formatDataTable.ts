type TsetTextTooltip = {
  text?: string;
  name?: string;
  maxDisplay?: number;
};

export const setTextTooltip = ({ text = '', name = '', maxDisplay = 0 }: TsetTextTooltip) =>
  !text || `${text}`.length <= maxDisplay || text === '' || !maxDisplay
    ? {
        [name]: {
          label: text,
        },
      }
    : {
        [name]: {
          label: `${text.substring(0, maxDisplay)}...`,
          hover: text,
        },
      };

type Tcolumn = {
  [key: string]: string;
};

export const setDisplay = (column: Tcolumn, maxDisplay?: number, setTextTooltipFn = setTextTooltip) =>
  Object.keys(column)?.map(keyColumn => setTextTooltipFn({ text: column[keyColumn], name: keyColumn, maxDisplay }))[0] ?? {};
