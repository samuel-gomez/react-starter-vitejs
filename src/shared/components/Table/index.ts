export { default as Body } from './Body';
export { default as Line } from './Body/Line';
export { default as Td } from './Body/Line/Td';
export { default as Headers } from './Headers';
export { default as Header } from './Headers/Header';
export { default as Th, type Torder } from './Headers/Header/Th';
export { default } from './Table.container';
export { ASCENDING, DESCENDING, NONE } from './constants';
export { setDisplay, setTextTooltip } from './formatDataTable';
