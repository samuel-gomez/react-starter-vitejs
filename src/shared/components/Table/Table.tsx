import { Table as TableTk } from '@axa-fr/react-toolkit-all';
import type { ComponentPropsWithoutRef, ReactNode } from 'react';
import Body, { type TBody } from './Body';
import Headers, { type THeaders } from './Headers';

export type TTable = THeaders &
  TBody & {
    childrenHeader?: ReactNode;
    className?: string;
    title: string;
    isCaptionVisible?: boolean;
  } & ComponentPropsWithoutRef<typeof TableTk>;

const Table = ({ onSort, sorting, headers, title, items, childrenHeader, isCaptionVisible = true, ...rest }: TTable) => (
  <TableTk {...rest}>
    <caption className={isCaptionVisible ? '' : 'sr-only'}>{title}</caption>
    <Headers headers={headers} onSort={onSort} sorting={sorting}>
      {childrenHeader}
    </Headers>
    <Body items={items} />
  </TableTk>
);

export default Table;
