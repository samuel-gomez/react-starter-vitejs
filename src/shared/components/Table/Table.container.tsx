import { ElementType, ReactNode } from 'react';
import { emptyFunction } from 'shared/helpers';
import Table, { TTable } from './Table';
import { DEFAULT_TABLE_ITEMS_TYPE } from './constants';

type TTableContainer = TTable & {
  children?: ReactNode;
  TableCmpt?: typeof Table;
  Fallback?: ElementType;
};

const TableContainer = ({
  children,
  TableCmpt = Table,
  items = [],
  headers = [],
  Fallback = emptyFunction,
  itemsType = DEFAULT_TABLE_ITEMS_TYPE,
  ...restTable
}: TTableContainer) =>
  items.length > 0 ? (
    <TableCmpt items={items} headers={headers} itemsType={itemsType} aria-label={`Tableau de ${itemsType}`} {...restTable}>
      {children}
    </TableCmpt>
  ) : (
    <Fallback />
  );
export default TableContainer;
