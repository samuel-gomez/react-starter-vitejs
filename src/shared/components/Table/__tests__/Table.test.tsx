import { render, screen, within } from 'shared/testsUtils/customRender';
import Table from '../Table';

const defaultProps = {
  classModifier: 'classModifier',
  title: 'my title',
  headers: [],
  items: [],
};

describe('Table', () => {
  it('Render <Table/> with empty items', () => {
    render(<Table {...defaultProps} />);
    const tbody = screen.getAllByRole('rowgroup')[1];
    expect(within(tbody).queryAllByRole('row').length).toBe(0);
  });

  const headers = [
    {
      label: 'label',
      key: 'uid',
      field: 'name',
    },
  ];
  const items = [
    {
      key: 'uid',
      cols: {
        name: {
          label: 'samuel',
        },
      },
    },
  ];

  it('Render <Table/> with 1 header and 1 item', () => {
    render(<Table {...defaultProps} headers={headers} items={items} />);

    const thead = screen.getAllByRole('rowgroup')[0];
    expect(within(thead).getAllByRole('columnheader').length).toBe(1);
    within(thead).getByText('label');

    const tbody = screen.getAllByRole('rowgroup')[1];
    expect(within(tbody).getAllByRole('row').length).toBe(1);
    within(tbody).getByText('samuel');
  });

  it('Render <Table/> without visible caption', async () => {
    render(<Table {...defaultProps} isCaptionVisible={false} />);
    expect(screen.getByText(defaultProps.title)).toHaveClass('sr-only');
  });
});
