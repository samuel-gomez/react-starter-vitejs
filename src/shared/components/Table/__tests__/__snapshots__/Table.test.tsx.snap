// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > Render <Table/> with 1 header and 1 item 1`] = `
<DocumentFragment>
  <table
    class="af-table af-table--classModifier"
  >
    <thead
      aria-label="table-header"
      class="af-table__thead"
    >
      <tr
        aria-label="table-header-line"
        class="af-table__tr"
      >
        <th
          class="af-table__th"
          role="columnheader"
          scope="col"
        >
          <span
            class="af-table__th-label"
          >
            label
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      aria-label="table-body"
      class="af-table__body"
    >
      <tr
        aria-label="table-body-line"
        class="af-table__tr"
      >
        <td
          class="af-table__cell"
        >
          samuel
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`Table > Render <Table/> with empty items 1`] = `
<DocumentFragment>
  <table
    class="af-table af-table--classModifier"
  >
    <thead
      aria-label="table-header"
      class="af-table__thead"
    >
      <tr
        aria-label="table-header-line"
        class="af-table__tr"
      />
    </thead>
    <tbody
      aria-label="table-body"
      class="af-table__body"
    />
  </table>
</DocumentFragment>
`;
